### PR TITLE
utils.service: add systemd options, make service calls safer

### PR DESCRIPTION
**Pull request by rdkdd**
- this PR solves https://trello.com/c/2kMETG2y/1327-make-avocadoutilsservice-thread-and-even-multi-access-safe?menu=filter&filter=serv
- furthermore are added `status, reset-failed, mask, unmask` options to service management (majority of which are transerred from https://github.com/avocado-framework/avocado-vt/blob/master/virttest/staging/service.py) - see https://github.com/avocado-framework/avocado-vt/pull/1565
- this is also bugfix for omission of `runner` parameter in `internal_servicemanagement`. Optional runners were then not possible.
- style correction of long lines

### CODE REVIEW

File: avocado/utils/service.py [hunk 1]
```diff
@@ -775,15 +775,12 @@ class has methods to start/stop services.
     # service_command_generator is explicitly undefined so that we get the
     # NameError on first access
     # pylint: disable=W0601
-    global _service_manager
```
Comment (ldoktor) [inline avocado/utils/service.py]: Would you please provide more details in the commit message? This is actually pretty serious bugfix so it deserves more description. I'd suggest describing the issue, the steps you did to address and perhaps you should mention it can affect the speed of calling "ServiceManager()" multiple times, but in return it fixes the issue. I agree that caching it doesn't make much sense as the actual call is fairly quick and we don't expect people creating ServiceManagers in loops.

File: avocado/utils/service.py [hunk 2]
```diff
@@ -277,6 +279,11 @@ def set_target_command(target):
             target = convert_systemd_target_to_runlevel(target)
             return ["telinit", target]
         return set_target_command
+    # Do not need reset failed, mask and unmask in sys_v style system.
+    elif command in ["reset_failed", "mask", "unmask"]:
```
Comment (ldoktor) [inline avocado/utils/service.py]: yep, these don't really map to systemv

File: avocado/utils/service.py [hunk 3]
```diff
@@ -267,6 +267,8 @@ def sys_v_init_command_generator(command):
     elif command == 'disable':
         command_name = "chkconfig"
         command = "off"
+    elif command == "raw_status":
```
Comment (ldoktor) [inline avocado/utils/service.py]: what is the `raw_status` command? I don't know it and I can't find it anywhere...

Comment (ldoktor) [reply avocado/utils/service.py]: Actually it maps to "status" on both. What is the purpose of this, it looks duplicate and unnecessary...

File: avocado/utils/service.py [hunk 4]
```diff
@@ -775,15 +775,12 @@ class has methods to start/stop services.
...
     # NameError on first access
     # pylint: disable=W0601
-    global _service_manager
-    try:
-        return _service_manager
-    except NameError:
-        internal_generator = _get_service_command_generator
-        internal_parser = _get_service_result_parser
-        _service_manager = internal_service_manager(internal_generator(run),
-                                                    internal_parser(run))
```
Comment (ldoktor) [inline avocado/utils/service.py]: nitpick: it'd be nice to split the bugfixes and first address the `run=run` issue along with description and then the caching issue as a separate commits. It might look unnecessary to you, but it makes it easier to follow and see what was actually fixed as well as it makes it actually possible to revert the actual little pieces should some problem arise.

**Comment by ldoktor** [comment]
Nice job, @rdkdd, would you please create a selftest for it? There are some in `selftests/unit/test_utils_service.py`.

File: avocado/utils/service.py (hunk 3 above)
Comment (rdkdd) [reply avocado/utils/service.py]: This is from https://github.com/avocado-framework/avocado-vt/blob/master/virttest/staging/service.py . It prints out complete status message of a process. I did not complete implementation. But it seems redundant here, so I will remove it.

File: avocado/utils/service.py (hunk 4 above)
Comment (ldoktor) [reply avocado/utils/service.py]: Thank you for the split. There is a small issue with the commit message as it claims this fixes the `multi-access` issue, which is not true. This only makes sure the runner is not omitted when creating the service class for the first time, the next commit fixes the multi-access issue.

Also please configure your git to wrap lines.

**Comment by ldoktor** [comment]
Actually I noticed the missing selftests, would you please address that? (Add selftest for 2 calls with different runner)

File: avocado/utils/service.py (hunk 4 above)
Comment (rdkdd) [reply avocado/utils/service.py]: @ldoktor IMHO commit message is ok. There is no `multi-access` word in it. Maybe you mixed it up with https://github.com/avocado-framework/avocado/pull/2640/commits/cec424f477357146c7a6074259cf7e075fd5b513. Will correct wrapping..

Comment (ldoktor) [reply avocado/utils/service.py]: When you apply this patch only and run the 2 commands mentioned in the card you claim it fixes it will fail. Therefor it is not addressing issue from that card...

Comment (ldoktor) [reply avocado/utils/service.py]: Well, simply add a selftest and you'll see ;-)

**Comment by rdkdd** [comment]
@ldoktor  could you please review this PR?
thanks

File: selftests/unit/test_utils_service.py [hunk 5]
```diff
@@ -26,6 +26,30 @@
...
 
+class TestRunCalls(unittest.TestCase):
+
+    def setUp(self):
+        self.run_param1 = ["foo_target", "set_target", mock.Mock()]
+        self.run_param2 = ["foo_service", "start", mock.Mock()]
+        self.run_param1[-1].return_value.stdout = "systemd"
+        self.run_param2[-1].return_value.stdout = "init"
+        self.results = ["systemctl isolate foo_target",
+                        "service foo_service start"]
```
Comment (ldoktor) [inline selftests/unit/test_utils_service.py]: Regarding this selftes, how many times do you expect this setup to be used? I mean calling `set_target` followed by `start` on 2 different instances... IMO this is way too specific to be a setup. The test itself is actually more generic than this setup, which is quite odd and makes it hard to follow. Unittests are meant to be simple, how about something like:


```python
class TestMultipleInstances(unittest.TestCase):

    def test_different_runners(self):
        # Call 'set_target' on first runner
        runner1 = mock.Mock()
        runner1.return_value.stdout = 'systemd'
        service1 = service.service_manager(run=runner1)
        service1.set_target('foo_target')
        self.assertEqual(runner1.call_args[0][0],
                         'systemctl isolate foo_target')
        # Call 'start' on second runner
        runner2 = mock.Mock()
        runner2.return_value.stdout = 'init'
        service2 = service.service_manager(run=runner2)
        service2.start('foo_service')
        self.assertEqual(runner2.call_args[0][0],
                         'service foo_service start')

```

I think it'd be easier to follow and see what is happening (yes, it's not that nice and it repeats patterns, but the readability is IMO way better, what do you think?)

File: avocado/utils/service.py [hunk 6]
```diff
@@ -316,6 +321,8 @@ def list_command(service_name):
         def set_target_command(target):
             return [command_name, "isolate", target]
         return set_target_command
+    elif command == "reset_failed":
+        command = "reset-failed"
```
Comment (ldoktor) [inline avocado/utils/service.py]: Oh yeah, this is because we can't name attributes `reset-failed`. Sure.

**Comment by ldoktor** [comment]
Merged a rebased version (without any changes). The style-fixes are comming up soon.
